### PR TITLE
feat(payroll): add the schedule-vs-clock audit panel

### PR DESCRIPTION
**Pull request by jdelgado2002**
## Summary

- Add a schedule-vs-clock audit to the Payroll page. It compares
  scheduled shifts with clock punches for the pay period.
- Flag five row types: `missing_clock`, `open_clock`, `time_mismatch`,
  `matched`, `unscheduled_clock`.
- Exclude draft shifts (`is_published: false`) from the audit. A draft
  shift is not a schedule the employee saw.
- Add a dialog so a manager can record a missed clock punch from a
  scheduled shift, in two clicks.
- Assign every clock session in a shift's window, not one session.
  This fixes an in-progress shift, which showed as `missing_clock`
  before its first punch closed.
- Assign a session to the shift with the largest overlap. Use the
  clock-in delta as the tie-break. An open session scores zero
  overlap for this rule; the clock-in delta then decides its shift.

## Rework: an in-row display, not a separate panel

The first version of this feature put the audit in its own panel,
above the payroll table. Design review found two problems: the panel
duplicated employee names already in the table below it, and a
manager had to scroll between the audit and the pay numbers for the
same employee.

The reworked design removes the separate panel. It replaces the
"Incomplete Time Punches Detected" alert and the old
`ScheduleClockAudit` component. The `Payroll.tsx` page now:

- Shows a `ClockAuditBar` inside the "Employee Payroll Details" card
  header. It holds the count chips (to fix, no clock-out, info,
  matched), a tolerance select, and a filter toggle per chip.
- Shows a row chip on each employee row in the payroll table, and a
  Regular Hrs gap subtext when scheduled and clocked hours differ.
- Expands a detail row under each employee (`EmployeeAuditDetail`)
  with one line per audit row: the scheduled and clocked times, the
  gap, the status, and a repair action for `missing_clock` and
  `open_clock` rows.
- Filters the table by the active chip. It hides the subtotal and
  grand total rows during a filter, and shows an `aria-live="polite"`
  note with the visible employee count.

One `RecordShiftClockDialog` at the page level serves every detail-row
action, per the single-dialog pattern in `CLAUDE.md`.

Design doc:
[docs/superpowers/specs/2026-08-18-payroll-audit-inline-design.md](https://github.com/toyiyo/nimble-pnl/blob/feature/payroll-clock-audit/docs/superpowers/specs/2026-08-18-payroll-audit-inline-design.md)

## Test plan

- `npm run typecheck`: clean.
- `npm run lint`: clean on every file this branch changed. Repo-wide
  `npm run lint` still shows a pre-existing baseline (1540 errors) in
  files this branch did not touch.
- `npm run test -- --run`: 775 test files (1 skipped), 9374 tests
  passed (2 skipped), 0 failures.
- `npm run test:db`: 2908 tests, 2908 passed, 0 failed, after a clean
  `npm run db:reset`.
- `npm run build`: succeeded. Pre-existing chunk-size warning only,
  not from this branch's files.
- `npm run test:e2e`: 227 tests, 214 passed, 12 skipped, 1 failed.
  The failure, `manual-sale-tip-not-doubled.spec.ts:58`, sits in a POS
  Sales file this branch never touched. Confirmed pre-existing on
  `main` by running the same spec in a disposable worktree at
  `origin/main`: same failure, same line.

## Known deferred review findings

- `src/utils/scheduleClockAudit.ts:259` — info — Performance review
  found no findings at confidence >= 80 for this branch's diff
  (`scheduleClockAudit.ts`, `useScheduleClockAudit.ts`,
  `ClockAuditBar.tsx`, `EmployeeAuditDetail.tsx`, `Payroll.tsx`). No
  missing `staleTime`, no unbounded waits, no unvirtualized 100+ item
  lists, no memory leaks, no sequential awaits to parallelize. —
  Deferred because it is informational only. No defect, no action
  needed.
- `src/utils/scheduleClockAudit.ts:244` — minor —
  `sessionShiftOverlapMinutes` re-parses `shift.start_time`/
  `end_time` and `session.clockIn`/`clockOut` into new `Date` objects
  for every (session, shift) candidate pair inside the nested
  assignment loop. This duplicates parses already done elsewhere for
  the same pair. — Deferred because the cost is negligible at
  realistic per-employee shift/session counts (10 to 20). The result
  is memoized with `useMemo` in `useScheduleClockAudit.ts`. Hoisting
  the `Date` parse out of the inner loop is not worth it unless input
  sizes become unbounded. The prior fold pass deferred this same
  finding for the same reason.

## Summary by CodeRabbit

* **New Features**
  * Added an integrated payroll schedule-versus-clock audit.
  * Added audit filters, tolerance selection, employee status indicators, expandable details, and repair actions.
  * Managers can enter missing or open clock data directly from payroll.
  * Added support for multiple clock sessions, gaps, in-progress shifts, and overnight schedules.
* **Bug Fixes**
  * Improved timezone handling, date-range accuracy, pagination, and incomplete-result detection.
* **Tests**
  * Added comprehensive unit and end-to-end coverage for audit calculations and payroll workflows.
* **Documentation**
  * Added design and implementation documentation for the payroll audit experience.

### CODE REVIEW

File: tests/e2e/schedule-clock-audit.spec.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+/** Create one hourly employee through the /scheduling UI. Returns the employee's name. */
+async function createHourlyEmployee(page: Page): Promise<string> {
+  await page.goto('/scheduling');
+  await expect(page.getByRole('heading', { name: /staff schedule/i })).toBeVisible({ timeout: 10000 });
+
+  await page.getByRole('button', { name: /employee/i }).first().click();
+  const dialog = page.getByRole('dialog', { name: /add new employee|edit employee/i });
+
+  const employeeName = `Audit Employee ${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/e2e/schedule-clock-audit.spec.ts]: **✅ Agreed** — Replaced Math.random with crypto.randomUUID for the unique test name, matching the existing codebase pattern. Fixed in `e8ecbf6d`.

File: src/hooks/useScheduleClockAudit.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  const endMs = end.getTime();
+
+  // The memo dependencies below are the primitive `startMs`/`endMs`, not the
+  // `start`/`end` Date objects: the caller (Payroll.tsx's `getDateRange`)
+  // rebuilds `start`/`end` as new Date instances on every render, which
+  // would otherwise bust every memo below on every unrelated page re-render.
+  const { fetchStart, fetchEnd } = useMemo(
+    () => bufferPunchFetchRange(start, end),
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- keyed on primitives, not the Date objects
+    [startMs, endMs],
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useScheduleClockAudit.ts]: **✅ Agreed** — Fixed. The hook now converts the viewer-local start/end into restaurant-zone instants via businessDayRangeToInstants, the usePayroll pattern, before it queries or filters. Fixed in `e8ecbf6d`.

File: src/utils/scheduleClockAudit.ts [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+  return sessions;
+}
+
+const sessionOverlapsShift = (session: WorkSession, shift: AuditShift, now: Date): boolean => {
+  const padMs = MATCH_PAD_MINUTES * MINUTE_MS;
+  const shiftStart = new Date(shift.start_time).getTime() - padMs;
+  const shiftEnd = new Date(shift.end_time).getTime() + padMs;
+  const sessionStart = new Date(session.clockIn).getTime();
+  const sessionEnd = session.clockOut ? new Date(session.clockOut).getTime() : now.getTime();
+  return sessionStart <= shiftEnd && sessionEnd >= shiftStart;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/utils/scheduleClockAudit.ts]: **✅ Agreed** — Fixed. An open session's clock-in must now land inside the shift's padded window; it no longer reaches forward to now, so a stale open session cannot match a later, unrelated shift. Fixed in `e8ecbf6d`.

File: src/hooks/useScheduleClockAudit.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    // shift that only got fetched because of this widened window, but does
+    // not actually overlap the period, is excluded.
+    queryKey: ['shifts', 'clock-audit', restaurantId, fetchStart.toISOString(), end.toISOString()],
+    queryFn: async () => {
+      if (!restaurantId) return [];
+      const { data, error } = await supabase
+        .from('shifts')
+        .select(
+          'id, restaurant_id, employee_id, start_time, end_time, break_duration, position, status, is_published',
+        )
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useScheduleClockAudit.ts]: **✅ Agreed** — Fixed. The shifts query now uses fetchAllRows with an id tiebreaker, matching the pagination pattern already used by usePayroll and useTimePunches. Fixed in `e8ecbf6d`.

File: src/components/payroll/ScheduleClockAudit.tsx [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+                >
+                  {t.label}
+                  <span className="ml-1.5 text-[11px] px-1.5 py-0.5 rounded-md bg-muted text-muted-foreground">
+                    {t.count}
+                  </span>
+                  {tab === t.key && (
+                    <div className="absolute bottom-0 left-0 right-0 h-[2px] bg-foreground" />
+                  )}
+                </button>
+              ))}
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/components/payroll/ScheduleClockAudit.tsx]: **✅ Agreed** — Fixed by switching to your suggested alternative: the filter control is now a plain button group with role=group and aria-pressed, not an incomplete ARIA tab pattern. Fixed in `e8ecbf6d`.

File: src/hooks/useScheduleClockAudit.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    [shifts, punches, startMs, endMs, toleranceMinutes],
+  );
+
+  return {
+    rows: result.rows,
+    summary: result.summary,
+    loading: shiftsLoading || punchesLoading,
+    error: shiftsError || punchesError || null,
+  };
+}
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useScheduleClockAudit.ts:126]: **✅ Agreed** — Added tests/unit/useScheduleClockAudit.test.ts: it covers the timezone normalization, the shifts query pagination, and the combined loading/error state. Fixed in `e8ecbf6d`.

File: src/utils/scheduleClockAudit.ts [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+    if (!session) {
+      // A shift that runs now with no punches is already a missed clock-in.
+      rows.push({
+        key: `shift-${shift.id}`,
+        status: 'missing_clock',
+        employeeId: shift.employee_id,
+        shift,
+        scheduledMinutes,
+      });
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/utils/scheduleClockAudit.ts]: **✅ Agreed** — Fixed to match the design doc: a shift that has not ended yet no longer counts as missing_clock. Added a regression test for a started-but-not-ended shift with no punches. Fixed in `e8ecbf6d`.

File: tests/e2e/schedule-clock-audit.spec.ts [hunk 8]
```diff
@@ -0,0 +1,195 @@
+/* eslint-disable @typescript-eslint/no-explicit-any -- window.__supabase and its test-only
+   helpers (exposeSupabaseHelpers) carry no type declarations, same as every other E2E spec. */
+import { test, expect, Page } from '@playwright/test';
+import { signUpAndCreateRestaurant, generateTestUser } from '../helpers/e2e-supabase';
+import { wallClockToInstant, formatLocalDateInTz } from '@/lib/shiftInterval';
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/e2e/schedule-clock-audit.spec.ts]: **✅ Agreed** — Fixed. The E2E spec now imports shiftInterval via a relative path, matching the E2E import convention (no @/ alias). Fixed in `e8ecbf6d`.

File: src/hooks/useScheduleClockAudit.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            'id, restaurant_id, employee_id, start_time, end_time, break_duration, position, status, is_published',
+          )
+          .eq('restaurant_id', restaurantId)
+          .gte('start_time', fetchStart.toISOString())
+          .lte('start_time', dayEnd.toISOString())
+          .order('start_time', { ascending: true })
+          .order('id')
+          .range(from, to),
+      );
+      return rows;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useScheduleClockAudit.ts]: **✅ Agreed** — Agreed. The hook now reads the capped flag from fetchAllRows. A capped result throws an error: the audit must not run on a truncated shift list, because it reports real shifts as absent. A unit test pins the behavior with 20 full pages. Fixed in `b0c32ca5`.

File: src/utils/scheduleClockAudit.ts [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+  );
+
+  // Nearest clock_in to the scheduled start wins.
+  candidates.sort(
+    (a, b) =>
+      Math.abs(minutesBetween(shift.start_time, a.clockIn)) -
+      Math.abs(minutesBetween(shift.start_time, b.clockIn)),
+  );
+  return candidates[0];
+};
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/utils/scheduleClockAudit.ts]: **✅ Agreed** — Agreed. Session assignment is now global: the audit collects all shift-session overlap pairs, sorts them by the absolute clock-in delta, and assigns the smallest delta first, one session per shift. A boundary session now pairs with the closest shift, not the first shift in time order. On the shift_id sub-point: AuditPunch has no shift_id field, and manager-entered punches match the shift times exactly, so their delta is 0 and the global ranking pairs them first. A unit test pins the back-to-back boundary case. Fixed in `b0c32ca5`.

File: src/components/payroll/ClockAuditBar.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+          !error &&
+          chips.map((chip) => (
+            <button
+              key={chip.filterClass}
+              type="button"
+              disabled={chip.count === 0}
+              aria-pressed={activeFilter === chip.filterClass}
+              onClick={() => toggleFilter(chip.filterClass)}
+              className={`text-[11px] px-2 py-1 rounded-md border font-medium transition-colors disabled:opacity-40 disabled:cursor-not-allowed ${chip.toneClass} ${
+                activeFilter === chip.filterClass ? 'ring-1 ring-foreground/40' : ''
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/components/payroll/ClockAuditBar.tsx:106]: **↩️ Pushed back** — The design doc sets this rule at line 181: a chip with count 0 renders disabled. The auto-clear effect in Payroll.tsx (from commit 7cb20c87) already drops a stuck filter when its count hits 0 or the audit query errors. No user can get stuck on a dead chip.

File: src/pages/Payroll.tsx [hunk 12]
```diff
@@ -438,6 +563,27 @@ const Payroll = () => {
...
+      return payrollGroups.map((group) => ({ ...group, totalCount: group.rows.length }));
+    }
+    return payrollGroups
+      .map((group) => ({
+        ...group,
+        rows: group.rows.filter((row) => employeeMatchesAuditFilter(auditRollup.get(row.employeeId), auditFilter)),
+        totalCount: group.rows.length,
+      }))
+      .filter((group) => group.rows.length > 0);
+  }, [payrollGroups, auditFilter, auditRollup]);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/Payroll.tsx]: **✅ Agreed** — Fixed. Payroll.tsx now derives appliedAuditFilter, null while the audit query loads or errors. Filtering, counts, and totals-row visibility read appliedAuditFilter, not the raw chip state. Fixed in `53e7a157`.

File: src/utils/scheduleClockAudit.ts [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+  const sessionsByShift = new Map<string, WorkSession[]>();
+  const matchedSessions = new Set<WorkSession>();
+  for (const pair of pairs) {
+    if (matchedSessions.has(pair.session)) continue;
+    const list = sessionsByShift.get(pair.shift.id) ?? [];
+    list.push(pair.session);
+    sessionsByShift.set(pair.shift.id, list);
+    matchedSessions.add(pair.session);
+  }
+  return { sessionsByShift, matchedSessions };
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/utils/scheduleClockAudit.ts]: **✅ Agreed** — Fixed. assignSessionsToShifts now checks a session's linked shift_id first, before the overlap and delta fallback. A regression test covers two shifts starting one minute apart. Fixed in `53e7a157`.

File: tests/e2e/schedule-clock-audit.spec.ts [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Step 6: activate the amber chip — the table shows only the flagged
+    // employee, and the totals row hides (a total over a filtered subset
+    // reads as a wrong pay total).
+    const totalRow = page.getByRole('rowheader', { name: 'TOTAL' });
+    await expect(totalRow).toBeVisible();
+    await toFixChip.click();
+    await expect(page.getByText('Clock filter active: 1 of 1 employees')).toBeVisible();
+    await expect(employeeRow).toBeVisible();
+    await expect(totalRow).toHaveCount(0);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/e2e/schedule-clock-audit.spec.ts:189]: **✅ Agreed** — Fixed. The e2e spec adds a second, unflagged employee (no shift, no punches). It stays visible unfiltered, drops out under the 1-of-2 to-fix filter, and returns once the filter clears. Fixed in `53e7a157`.

File: tests/unit/scheduleClockAudit.test.ts [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+      [
+        shift({ id: 'a', start_time: '2026-08-12T12:00:00Z', end_time: '2026-08-12T12:35:00Z' }),
+        shift({ id: 'b', start_time: '2026-08-12T16:00:00Z', end_time: '2026-08-12T22:00:00Z' }),
+      ],
+      [punch('emp1', 'clock_in', '2026-08-12T12:30:00Z')],
+    );
+    const byKey = new Map(result.rows.map((row) => [row.key, row]));
+    expect(byKey.get('shift-a')?.status).toBe('open_clock');
+    expect(byKey.get('shift-b')?.status).toBe('missing_clock');
+  });
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/unit/scheduleClockAudit.test.ts]: **✅ Agreed** — Fixed. Both test names now start with should, matching the file's naming convention. Fixed in `53e7a157`.
